apps/champions/lib/champions/units.ex

```diff
@@ -207,6 +217,151 @@ defmodule Champions.Units do
+
+      {:consumed_units_count, false} ->
+        {:error, :consumed_units_not_found}
+
+      {:consumed_units_valid, false} ->
+        {:error, :consumed_units_invalid}
+    end
+  end
+
+  defp delete_consumed_units(unit_ids) do
+    {amount_deleted, _return} = Units.delete_units(unit_ids)
+
+    if Enum.count(unit_ids) == amount_deleted, do: {:ok, amount_deleted}, else: {:error, "failed"}
+  end
+
+  defp validate_consumed_units(unit, unit_list) do
+    {same_character_amount, same_character_rank} = same_character_requirements(unit)
+    {same_faction_amount, same_faction_rank} = same_faction_requirements(unit)
+
+    try do
+      validate_units(
+        unit,
+        unit_list,
+        same_character_amount,
+        same_character_rank,
+        same_faction_amount,
+        same_faction_rank
+      )
+    rescue
+      _e in RuntimeError -> false
+    end
+  end
+
+  defp same_character_requirements(%Unit{rank: @star4}), do: {2, @star4}
+  defp same_character_requirements(%Unit{rank: @star5}), do: {1, @star5}
+  defp same_character_requirements(%Unit{rank: @illumination1}), do: {1, @star5}
+  defp same_character_requirements(%Unit{rank: @illumination2}), do: {1, @star5}
+  defp same_character_requirements(%Unit{rank: @illumination3}), do: {3, @star5}
+
+  defp same_faction_requirements(%Unit{rank: @star4}), do: {4, @star4}
+  defp same_faction_requirements(%Unit{rank: @star5}), do: {4, @star5}
+  defp same_faction_requirements(%Unit{rank: @illumination1}), do: {1, @illumination1}
+  defp same_faction_requirements(%Unit{rank: @illumination2}), do: {2, @illumination2}
+  defp same_faction_requirements(%Unit{rank: @illumination3}), do: {2, @illumination2}
+
+  defp validate_units(
+         unit,
+         unit_list,
+         same_character_amount,
+         same_character_rank,
+         same_faction_amount,
+         same_faction_rank
+       ) do
+    # Remove necessary units with the same character
+    removed_same_character =
+      Enum.reduce(1..same_character_amount, unit_list, fn _, list ->
+        same_character =
+          Enum.find(
+            list,
+            &(&1.character_id == unit.character_id and &1.rank == same_character_rank)
+          )
+
+        if is_nil(same_character), do: raise("not enough of same character")
+        if same_character.user_id != unit.user_id, do: raise("consumed unit not owned")
+
+        List.delete(list, same_character)
+      end)
+
+    # Remove necessary units with the same faction
+    removed_same_faction =
+      Enum.reduce(1..same_faction_amount, removed_same_character, fn _, list ->
+        same_faction =
+          Enum.find(
+            list,
+            &(&1.character.faction == unit.character.faction and &1.rank == same_faction_rank)
+          )
+
+        if is_nil(same_faction), do: raise("not enough of same faction")
+        if same_faction.user_id != unit.user_id, do: raise("consumed unit not owned")
+
+        List.delete(list, same_faction)
+      end)
```

Comment:
I'm not a huge fan of this approach of raising to indicates that the operation was not succefully, why not using the same approach as fuse to check this? if the objective is just to return a boolean when the user have enough units it should be straightfoward

```
  defp validate_consumed_units(unit, unit_list) do
    with {:removed_same_unit, {:ok, removed_same_unit}} <-
           {:removed_same_unit,
            remove_same_unit(unit, unit_list, same_character_requirements(unit))},
         {:removed_same_faction, {:ok, _removed_same_faction}} <-
           {:removed_same_faction,
            remove_same_faction(unit, removed_same_unit, same_faction_requirements(unit))} do
      true
    else
      _ -> false
    end
  end
```

And the two removals would be:

```

  defp remove_same_unit(unit, unit_list, {same_character_amount, same_character_rank}) do
    same_character_list =
      Enum.count(
        unit_list,
        &(&1.character_id == unit.character_id and &1.rank == same_character_rank)
      )

    if Enum.count(same_character_list) >= same_character_amount do
      deleted_same_character =
        Enum.reduce(same_character_list, unit_list, fn same_character, list ->
          List.delete(list, same_character)
        end)

      {:ok, deleted_same_character}
    else
      {:error, :not_enough_same_character}
    end
  end

  defp remove_same_faction(unit, unit_list, {same_faction_amount, same_faction_rank}) do
    same_faction_list =
      Enum.count(
        unit_list,
        &(&1.character.faction == unit.character.faction and &1.rank == same_faction_rank)
      )

    if Enum.count(same_faction_list) >= same_faction_amount do
      deleted_same_faction =
        Enum.reduce(same_faction_list, unit_list, fn same_faction, list ->
          List.delete(list, same_faction)
        end)

      {:ok, deleted_same_faction}
    else
      {:error, :not_enough_same_faction}
    end
  end
```

What do you think?

Reply:
The idea behind the raises was to stop execution early whenever we found something that was wrong. Is there something inherently bad about raises?

Nevertheless I've taken your approach and fixed it up a little bit, let me know if you think [this](https://github.com/lambdaclass/mirra_backend/pull/332/commits/68028f1770aee879ffb59cf34b23097220bb7836) works and is clearer than what was previously implemented!

Reply:
hmm i don't like using raises to stop execution because we won't know if the stop was to the player not having enough champions or we have something wrong in the code, for example we change the name of the character in units from `unit.character` to `unit.chara`, the code will rise when we try to do `unit.character.faction` but we'll never know if the rising was due to the code failing or by our check stoping execution

I you wan't to achieve the execution stopping earlier i think there're better approaches like using recursion or `Enum.reduce_while/3`

Reply:
Right, or we could have caught the exact custom errors we were raising in our rescue too. I've implemented it with reduce_while like you said!

Reply:
right! yes, handling specific errors could be another way to interrump execution, but i'd rater let the program crash when it should and stop the execution when we wan't explicitly. Maybe you can made my change my mind